Add E2E Cachi2 test case

lkolacek opened the pull request:
Test case checks fetching retrodep dependencies, generating environment vars file,
building image with all prepared prerequisites and printing help message for retrodep
app in built image

CLOUDBLD-12632

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] ~Docs updated (if applicable)~
- [ ] ~Docs links in the code are still valid (if docs were updated)~

lkolacek commented:
PTAL if you agree with this approach.
I haven't updated retrodep source repository I just used go1.18 in test build [Containerfile](https://github.com/containerbuildsystem/cachi2/pull/61/files#diff-0e7a49653d83029b114bcea7d9c486ca75db7064f3ddb93cc45508f2acf42cacR1). I am not sure if this is sufficient enough or more changes are required. 

But Cachi2 seems to be working fine!

lkolacek commented on tests/integration/test_package_managers.py:
```diff
@@ -161,3 +182,24 @@ def test_packages(
         log.info("Compare checksums of files in source vendor folder")
         assert files_checksums == expected_files_checksums
+
+    if test_params.build_image_with_fetched_deps:
```
I can try to break tests to some smaller separate pytests, if you want.

chmeliik replied:
Yeah I would prefer for the e2e tests to be separate

lkolacek replied:
Done

chmeliik commented on tests/integration/test_data/gomod_e2e_test/container/Containerfile:
```diff
@@ -0,0 +1,16 @@
+WORKDIR /tmp/gomod_e2e_test-source
+RUN . /tmp/cachi2.env && go build -o /retrodep
+
+FROM registry.access.redhat.com/ubi9/ubi-minimal:9.0.0
+
+COPY --from=build /retrodep /usr/bin/retrodep
```
IMO we don't need a second stage, single stage should be fine to enough that the build works (and should be faster since we won't need to pull a second base image)

lkolacek replied:
Ok changed to single stage build

chmeliik commented on tests/integration/test_data/gomod_e2e_test/container/Containerfile:
```diff
@@ -0,0 +1,16 @@
+
+COPY --from=build /retrodep /usr/bin/retrodep
+
+CMD ls | retrodep
```
Maybe `retrodep --help` would be clearer? I don't think piping `ls` output to retrodep does anything reasonable

lkolacek replied:
Changed.

chmeliik commented on tests/integration/test_package_managers.py:
```diff
@@ -161,3 +182,24 @@ def test_packages(
+            f"on built image {test_image.repository}"
+        )
+        (output, rc) = test_image.run_cmd_on_image([test_params.install_app, "--help"], tmpdir)
+        assert rc == 0, (
+            f"{test_params.install_app} was not installed on image." f"output-cmd: {output}"
+        )
+        test_image.delete_image()
```
If the assertion fails, the image will not be deleted. Maybe ContainerImage could be a context manager and delete the image on `__exit__`?

```python
with utils.build_image(...) as test_image:
    assert ...
```
Or a `try: assert ... finally: delete` would also work

lkolacek replied:
Great idea, thank you.  :taco:   Changed, please check. :)

chmeliik commented on tests/integration/utils.py:
```diff
@@ -42,6 +44,30 @@ def run_cmd_on_image(self, cmd: List, tmpdir: Path) -> Tuple[str, int]:
+
+
+def build_image(tmpdir: Path, container_folder: str, test_case: str) -> ContainerImage:
+    image_cmd = [
+        "podman",
+        "build",
+        container_folder,
+        "-v",
+        f"{tmpdir}:/tmp:z",
+        "--network",
+        "none",
+        "--tag",
+        test_case,
+    ]
```
Should we build with `--no-cache` to avoid succeeding accidentally?

lkolacek replied:
yes, added

chmeliik commented:
Overall this way of e2e testing does make sense to me :+1:

chmeliik commented:
I haven't updated retrodep source repository I just used go1.18 in test build [Containerfile](https://github.com/containerbuildsystem/cachi2/pull/61/files#diff-0e7a49653d83029b114bcea7d9c486ca75db7064f3ddb93cc45508f2acf42cacR1). I am not sure if this is sufficient enough or more changes are required.

I would update retrodep as well. I'm pretty sure the go.mod file will change when it gets processed with go 1.18

lkolacek commented:
> > I haven't updated retrodep source repository I just used go1.18 in test build [Containerfile](https://github.com/containerbuildsystem/cachi2/pull/61/files#diff-0e7a49653d83029b114bcea7d9c486ca75db7064f3ddb93cc45508f2acf42cacR1). I am not sure if this is sufficient enough or more changes are required.
> 
> I would update retrodep as well. I'm pretty sure the go.mod file will change when it gets processed with go 1.18

ACK, PR opened. Once it's merged, I'll update commit reference for our test.
* https://github.com/cachito-testing/retrodep/pull/1

brunoapimentel commented on tests/integration/test_package_managers.py:
```diff
@@ -161,3 +123,66 @@ def test_packages(
+
+    log.info("Create cachi2.env file")
+    env_vars_file = os.path.join(tmpdir, "cachi2.env")
+    cmd = ["generate-env", output_folder, "--output", env_vars_file, "--for-output-dir", "/tmp"]
```
I think the output-dir is incorrect here, it should be `/tmp/gomod_e2e_test-output`.

brunoapimentel commented:
I don't think the network isolation is working properly in the Gihub Actions environment. When I try to run the e2e test locally I get errors with Go trying to access the network, but it is passing in CI right now.

First issue  I can see is with the Buildah version, it's 1.23, which means the bug that affects isolation is not fixed yet:

https://github.com/actions/runner-images/blob/ubuntu22/20221212.1/images/linux/Ubuntu2204-Readme.md?plain=1#L68

The other one is that the env variables are not pointing to the right folder (see my comment [here](https://github.com/containerbuildsystem/cachi2/pull/61/files#r1063735326))

lkolacek commented:
@brunoapimentel @chmeliik please take a look at this approach.

1.) I failed with using the fedora container - building the cachi2 image stopped working.
2.) Then I failed with creating and using isolated-network as mentioned in usage Doc. 
* https://github.com/containerbuildsystem/cachi2/actions/runs/3890928216/jobs/6640555833

3.) And for the last try I picked building buildah with go as you can see here.
* It has 1 big disadvantage. After this installation buildah and podman do not share registries anymore and `podman images` and `buildah images` return different lists. That's why I adjusted the cleanup method to use the buildah command instead of podman.

I added a test if network isolation works this way. I believe it might be a sufficient workaround for us for now. If you want I can divide tests to run all except E2E ones with old buildah and just for E2e use the built one. 

What do you think?

chmeliik commented:
> 1.) I failed with using the fedora container - building the cachi2 image stopped working.

What was the issue?

lkolacek commented:
@taylormadore @brunoapimentel @chmeliik PR should be ready for final review.

taylormadore commented on tests/integration/test_data/gomod_e2e_test/container/Containerfile-fail:
```diff
@@ -0,0 +1,15 @@
+FROM docker.io/golang:1.18.1 AS build
+
+# Network access should be disabled for testing Cachi2 functionality
+RUN curl www.google.com
```
When testing locally, I was getting this error when the build had network access:
> `FAILED tests/integration/test_package_managers.py::test_e2e[gomod_e2e_test] - UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 10810: invalid start byte`

Maybe we could just fetch the header and suppress the other output? Something like `curl -IsS www.google.com`?

chmeliik replied:
+1

Also, do we need a second Containerfile for this? How about checking it in the successful build:
```Dockerfile
RUN if curl -IsS www.google.com; then echo "Has network access!"; exit 1; fi
```

lkolacek replied:
Great idea. changed

brunoapimentel commented on .github/workflows/gating.yaml:
```diff
@@ -104,17 +104,22 @@ jobs:
   build-image:
     name: Build Cachi2 image and run smoketest on it
```
Since we're not running the smoketest anymore, we should rename this.

lkolacek replied:
renamed

brunoapimentel commented on tests/integration/test_package_managers.py:
```diff
@@ -160,3 +123,82 @@ def test_packages(
+        "--for-output-dir",
+        os.path.join("/tmp", f"{test_case}-output"),
+    ]
+    (output, rc) = cachi2_image.run_cmd_on_image(cmd, tmpdir)
```
Nitpick: I don't like the abbreviation here, it took me a minute to figure out what `rc` stands for.

lkolacek replied:
Added extra refactoring commit for this :)

chmeliik commented on tests/integration/test_package_managers.py:
```diff
@@ -111,45 +110,9 @@ def test_packages(
-        log.info("Compare checksums of fetched deps files")
-        assert files_checksums == expected_files_checksums
 
     if test_params.check_vendor_checksums:
```
Shouldn't the vendor check also be part of `fetch_deps_and_check_output()`?

lkolacek replied:
sure, moved to fetch_deps_and_check_output

chmeliik commented on tests/integration/test_package_managers.py:
```diff
@@ -160,3 +123,82 @@ def test_packages(
+                ref="c3496edd5d45523a1ed300de1575a212b86d00d3",
+                packages=({"path": ".", "type": "gomod"},),
+                check_vendor_checksums=False,
+                install_app="retrodep",
```
two nitpicks for `install_app`

1. Should it be a separate pytest param, not part of TestParameters? The command is never relevant for non-e2e tests
2. Could it be something like a `check_cmd: list[str]` instead (and set it to `["retrodep", "--help"]`)? Not all e2e tests will want to call `<some_app> --help`

lkolacek replied:
ack great idea

chmeliik commented:
> Run GitHub action pipeline in Fedora container
>
> Remove simple smoketest step from pipeline since it is covered in integration tests

These should be two separate commits

lkolacek commented on tests/integration/utils.py:
```diff
@@ -42,6 +46,38 @@ def run_cmd_on_image(self, cmd: List, tmpdir: Path) -> Tuple[str, int]:
+        image_cmd = ["podman", "rmi", "--force", self.repository]
+        (output, rc) = run_cmd(image_cmd)
+        if rc != 0:
+            raise RuntimeError(f"Image deletion failed. Output:{output}")
```
ou removed

lkolacek commented on tests/integration/utils.py:
```diff
@@ -42,6 +46,38 @@ def run_cmd_on_image(self, cmd: List, tmpdir: Path) -> Tuple[str, int]:
+        "-f",
+        containerfile,
+        "-v",
+        f"{tmpdir}:/tmp:z",
```
Fixed. sorry I work on ubuntu and sometimes I receive different errors.

lkolacek commented:
> > 1.) I failed with using the fedora container - building the cachi2 image stopped working.
> 
> What was the issue?

Running fedora container in privileged mode and changing order of steps for git checkout.

lkolacek commented:
I hope I addressed everything. Big thanks for all feedback @taylormadore @chmeliik @brunoapimentel PTAL

brunoapimentel commented on tests/integration/test_package_managers.py:
```diff
@@ -147,52 +147,74 @@ def test_packages(
+    cachi2_image: utils.ContainerImage,
+    tmpdir: Path,
+    test_data_dir: Path,
+    request,
```
Just a minor thing, but we can annotate this with `pytest.FixtureRequest`

lkolacek replied:
Done

brunoapimentel reviewed (approved):
Thanks for fighting bravely with this, Lada. The e2e test is looking awesome! :rocket: :rocket: :rocket:

lkolacek commented on tests/integration/test_package_managers.py:
```diff
@@ -147,52 +147,74 @@ def test_packages(
+    request: pytest.FixtureRequest,
+):
+    """
+    End to end test for package manages.
```
fixed
